tests/unit/s2n_server_hello_retry_test.c

```diff
@@ -56,6 +106,118 @@ static int client_hello_detect_duplicate_calls(struct s2n_connection *conn, void
+    EXPECT_SUCCESS(s2n_negotiate(conn, &blocked));
+    EXPECT_EQUAL(ch_ctx->invocations, 3);
+
+    return S2N_SUCCESS;
+}
+
+S2N_RESULT hello_retry_client_hello_cb_test(bool enable_poll) {
+        struct s2n_config *server_config;
+        struct s2n_config *client_config;
+
+        struct s2n_connection *server_conn;
+        struct s2n_connection *client_conn;
+
+        struct s2n_cert_chain_and_key *tls13_chain_and_key;
+
+        EXPECT_NOT_NULL(server_config = s2n_config_new());
+        EXPECT_NOT_NULL(client_config = s2n_config_new());
+
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(client_config, "default_tls13"));
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(server_config, "default_tls13"));
+
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(server_config));
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(client_config));
+
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&tls13_chain_and_key,
+                S2N_ECDSA_P384_PKCS1_CERT_CHAIN, S2N_ECDSA_P384_PKCS1_KEY));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(client_config, tls13_chain_and_key));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, tls13_chain_and_key));
+
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, client_config));
+
+        struct s2n_test_io_pair io_pair;
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Force HRR path */
+        client_conn->security_policy_override = &security_policy_test_tls13_retry;
+
+        /* setup the client hello callback */
+        struct client_hello_context client_hello_ctx = {.invocations = 0,
+            .mode = S2N_CLIENT_HELLO_CB_NONBLOCKING, .mark_done = false,
+            .enable_poll = enable_poll };
+        EXPECT_SUCCESS(s2n_config_set_client_hello_cb(server_config,
+                s2n_client_hello_poll_cb, &client_hello_ctx));
+        EXPECT_SUCCESS(s2n_config_set_client_hello_cb_mode(server_config,
+            S2N_CLIENT_HELLO_CB_NONBLOCKING));
+
+        if (enable_poll) {
+            /* Enable callback polling mode */
+            EXPECT_SUCCESS(s2n_config_client_hello_cb_enable_poll(server_config));
+        }
+
+        /* negotiate and make assertions */
+        EXPECT_SUCCESS(s2n_negotiate_poll_hello_retry(server_conn, client_conn, &client_hello_ctx));
+
+        /* check hello retry state */
+        EXPECT_TRUE(server_conn->handshake.handshake_type & HELLO_RETRY_REQUEST);
+        EXPECT_TRUE(IS_HELLO_RETRY_HANDSHAKE(client_conn));
+        EXPECT_TRUE(IS_HELLO_RETRY_HANDSHAKE(server_conn));
```

Comment:
You don't need the first manual check, and it could break your test someday if we change the implementation. Just use the macro.